keadm/cmd/keadm/app/cmd/edge/rollback.go

```diff
@@ -0,0 +1,113 @@
+limitations under the License.
+*/
+
+package edge
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kubeedge/kubeedge/common/constants"
+	"github.com/kubeedge/kubeedge/edge/pkg/common/util"
+	"github.com/kubeedge/kubeedge/pkg/apis/componentconfig/edgecore/v1alpha2"
+	api "github.com/kubeedge/kubeedge/pkg/apis/fsm/v1alpha1"
+	"github.com/kubeedge/kubeedge/pkg/util/fsm"
+	"github.com/kubeedge/kubeedge/pkg/version"
+)
+
+// NewEdgeUpgrade returns KubeEdge edge upgrade command.
+func NewEdgeRollback() *cobra.Command {
+	rollbackOptions := newRollbackOptions()
+
+	cmd := &cobra.Command{
+		Use:   "rollback",
+		Short: "rollback edge component. Rollback the edge node to the desired version.",
+		Long:  "Rollback edge component. Rollback the edge node to the desired version.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// upgrade edgecore
+			return rollbackOptions.rollback()
+		},
+	}
+
+	addRollbackFlags(cmd, rollbackOptions)
+	return cmd
+}
+
+// newJoinOptions returns a struct ready for being used for creating cmd join flags.
+func newRollbackOptions() *RollbackOptions {
+	opts := &RollbackOptions{}
+	opts.HistoryVersion = version.Get().String()
+	opts.Config = constants.DefaultConfigDir + "edgecore.yaml"
+
+	return opts
+}
+
+func (ro *RollbackOptions) rollback() error {
+	// get EdgeCore configuration from edgecore.yaml config file
+	data, err := os.ReadFile(ro.Config)
+	if err != nil {
+		return fmt.Errorf("failed to read config file %s: %v", ro.Config, err)
+	}
+
+	configure := &v1alpha2.EdgeCoreConfig{}
+	err = yaml.Unmarshal(data, configure)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal config file %s: %v", ro.Config, err)
+	}
+	event := &fsm.Event{
+		Type:   "RollBack",
+		Action: api.ActionSuccess,
+	}
+	defer func() {
+		// report upgrade result to cloudhub
+		if err = util.ReportUpgradeResult(configure, ro.TaskType, ro.TaskName, *event); err != nil {
+			klog.Warning("failed to report upgrade result to cloud: %v", err)
+		}
+	}()
+
+	rbErr := rollback(ro.HistoryVersion, configure.DataBase.DataSource, ro.Config)
+	if rbErr != nil {
+		event.Action = api.ActionFailure
+		event.ErrorMsg = fmt.Sprintf("upgrade error: %v, rollback error: %v", err, rbErr)
+	}
+
+	return nil
+}
+
+type RollbackOptions struct {
+	HistoryVersion string
+	TaskType       string
+	TaskName       string
+	Config         string
+}
+
+func addRollbackFlags(cmd *cobra.Command, rollbackOptions *RollbackOptions) {
+	cmd.Flags().StringVar(&rollbackOptions.HistoryVersion, "history", rollbackOptions.HistoryVersion,
+		"Use this key to specify the origin version before upgrade")
+
+	cmd.Flags().StringVar(&rollbackOptions.Config, "config", rollbackOptions.Config,
+		"Use this key to specify the path to the edgecore configuration file.")
+
+	cmd.Flags().StringVar(&rollbackOptions.TaskType, "type", rollbackOptions.Config,
```

Comment:
The default value should be `rollbackOptions.TaskType`, what enumeration does the task type have, and will it be a constant value?

Reply:
Yes. It should be `Rollback`.